File: C++/0005_longest_palindromic_substr.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if(s.length() > 1000 || s.length() == 0){ return ""; }
        if(s.length() == 1){ return s; }
        if(s.length() == 2){ return s[0] == s[1] ? s : s.substr(0, 1); }
        
        string rt = "";
        int maxlen = -1;
        for(int m = 1; m < s.length(); ++m){
            int i = m - 1, j = m + 1;
            while(i >= 0 && j < s.length() && s[i] == s[j]){
                --i;
                ++j;
            }
            if(maxlen < j - i - 1){
                maxlen = j - i - 1;
                rt = s.substr(i + 1, maxlen);
            } 
            i = m - 1;
            j = m;
            while(i >= 0 && j < s.length() && s[i] == s[j]){
                --i;
                ++j;
            }
             if(maxlen < j - i - 1){
                maxlen = j - i - 1;
                rt = s.substr(i + 1, maxlen);
            } 
        }
        return rt;
    }
};

File: Python/0005_longest_palindromic_substr.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if(len(s) > 1000 or len(s) <= 0):
            return None
        if(1 == len(s)):
            return s
        if(2 == len(s)):
            if s[0] == s[1]:
                return s
            else:
                return s[0]
        maxlen = 1
        rt = s[0]
        for m in range(1, len(s)):
            i = m - 1
            j = m + 1
            rt, maxlen = self.checkpalingdrome(i, j, s, rt, maxlen)
            
            i = m - 1
            j = m
            rt, maxlen = self.checkpalingdrome(i, j, s, rt, maxlen)
        return rt
        
    def checkpalingdrome(self, i, j, s, rt, maxlen):
        while(i >= 0 and j < len(s) and s[i] == s[j]):
            i -= 1
            j += 1
        if(maxlen < j - i - 1):
            maxlen = j - i - 1
            rt = s[i+1:j]
        return rt, maxlen
